### PR TITLE
WIP: Initial FreeBSD port

### PR DESCRIPTION
**Pull request by mkopec**
This patch enables initial support for FreeBSD.

Currently it does not compile due to fwupd/fwupd#3307

### CODE REVIEW

**Comment by mkopec** [comment]
This branch should now compile successfully. ~However, it requires efivar with https://github.com/rhboot/efivar/commit/4af5b56066b4af66ce8d968948bb2bf4f5a87c79 applied.~

The actual fix was to uninstall devel/efivar from ports and use the one provided by FreeBSD.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+guint64
+fu_efivar_space_used (GError **error)
+{
+	guint64 total = 0;
+	efi_guid_t *guid = NULL;
+	char *name = NULL;
+
+	g_return_val_if_fail (error == NULL || *error == NULL, G_MAXUINT64);
+
+	while (efi_get_next_variable_name(&guid, &name)) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+ * Returns: %TRUE on success
+ *
+ * Since: 1.4.0
+ **/
+gboolean
+fu_efivar_set_data (const gchar *guids, const gchar *name, const guint8 *data,
+		     gsize sz, guint32 attr, GError **error)
+{
+	efi_guid_t guid;
+	efi_str_to_guid(guids, &guid);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+	if (efi_set_variable(guid, name, data, sz, attr) != 0) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+ *
+ * Returns: %TRUE on success
+ *
+ * Since: 1.4.0
+ **/
+gboolean
+fu_efivar_delete (const gchar *guids, const gchar *name, GError **error)
+{
+	efi_guid_t guid;
+	efi_str_to_guid(guids, &guid);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+ *
+ * Returns: %TRUE on success
+ *
+ * Since: 1.4.0
+ **/
+gboolean
+fu_efivar_delete_with_glob (const gchar *guids, const gchar *name_glob, GError **error)
+{
+	efi_guid_t guid_to_delete;
+	efi_str_to_guid(guids, &guid_to_delete);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+	efi_guid_t *guid = NULL;
+	gchar *name = NULL;
+
+	gboolean rv = FALSE;
+
+	while (efi_get_next_variable_name (&guid, &name)) {
+		if (memcmp(&guid_to_delete, guid, sizeof(guid)) == 0) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+			if (fu_common_fnmatch (name, name_glob)) {
+				rv = fu_efivar_delete(guids, name, error);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+	return rv;
+}
+
+static gboolean
+fu_efivar_exists_guid (const gchar *guids)
+{
+	efi_guid_t *guid = NULL;
+	gchar *name = NULL;
+	efi_guid_t test;
+	efi_str_to_guid(guids, &test);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+
+	while (efi_get_next_variable_name (&guid, &name)) {
+		if (memcmp(&test, guid, sizeof(test)) == 0) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+fu_efivar_get_names (const gchar *guids, GError **error)
+{
+	const gchar *name_guid;
+	g_autoptr(GPtrArray) names = g_ptr_array_new_with_free_func (g_free);
+
+	/* find names with matching GUID */
+	efi_guid_t *guid = NULL;
+	gchar *name = NULL;
+	efi_guid_t test;
+	efi_str_to_guid(guids, &test);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+	while (efi_get_next_variable_name (&guid, &name)) {
+		if (memcmp(&test, guid, sizeof(test)) == 0) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please, add space after function name.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+/**
+ * fu_efivar_space_used:
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the total size used by all EFI variables. This may be less than the size reported by the
+ * kernel as some (hopefully small) variables are hidden from userspace.
+ *
+ * Returns: total allocated size of all visible variables, or %G_MAXUINT64 on error
+ *
+ * Since: 1.5.1
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please change version tags.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 13]
```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2021 Norbert Kamiński <norbert.kaminski@3mdeb.com>
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c:3]: Please, add yourself and Sergii

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+#include <fcntl.h>
+#include <string.h>
+#include <efivar.h>
+
+#include "fu-common.h"
+#include "fu-efivar.h"
+
+#include "fwupd-error.h"
+
+static gchar *
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c:28]: Please, add function description

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+ * Returns: %TRUE on success
+ *
+ * Since: 1.4.0
+ **/
+gboolean
+fu_efivar_get_data (const gchar *guids, const gchar *name, guint8 **data,
+		       gsize *data_sz, guint32 *attr, GError **error)
+{
+	efi_guid_t guid;
+	efi_str_to_guid(guids, &guid);
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Please add space function name.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 11 above)
Comment (Asiderr) [reply libfwupdplugin/fu-efivar-freebsd.c]: Brackets not needed.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+{
+	guint64 total = 0;
+	efi_guid_t *guid = NULL;
+	char *name = NULL;
+
+	g_return_val_if_fail (error == NULL || *error == NULL, G_MAXUINT64);
+
+	while (efi_get_next_variable_name(&guid, &name)) {
+		size_t size = 0;
+		if (efi_get_variable_size (*guid, name, &size) < 0) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Brackets not needed.

File: libfwupdplugin/fu-efivar-freebsd.c [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+	efi_guid_t *guid = NULL;
+	gchar *name = NULL;
+
+	gboolean rv = FALSE;
+
+	while (efi_get_next_variable_name (&guid, &name)) {
+		if (memcmp(&guid_to_delete, guid, sizeof(guid)) == 0) {
+			if (fu_common_fnmatch (name, name_glob)) {
+				rv = fu_efivar_delete(guids, name, error);
+				if (!rv) {
```
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Brackets not needed.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 6 above)
Comment (Asiderr) [inline libfwupdplugin/fu-efivar-freebsd.c]: Avoid nested `if`s. Use reversed conditions e.g.:
```
if (memcmp(&guid_to_delete, guid, sizeof(guid)) != 0)
        continue;

if (!fu_common_fnmatch (name, name_glob))
        continue;

rv = fu_efivar_delete(guids, name, error);
if (!rv)
        break;
```

**Review by Asiderr — CHANGES_REQUESTED** [review]

File: libfwupdplugin/meson.build [hunk 18]
```diff
@@ -125,12 +124,19 @@ library_deps = [
...
   libjcat,
   platform_deps,
+  efivar,
 ]
 
 if get_option('libarchive')
   library_deps += libarchive
 endif
 
+if host_machine.system() == 'linux'
```
Comment (Asiderr) [inline libfwupdplugin/meson.build]: If freebsd - `fu-efivar-freebsd.c`, else `fu-efivar.c`. Windows also will use `fu-efivar.c`.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 1 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 2 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 3 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 4 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 5 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 6 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 7 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 8 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 9 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 10 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 11 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 12 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 13 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c:3]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 14 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c:28]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 15 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 16 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 17 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

File: libfwupdplugin/meson.build (hunk 18 above)
Comment (mkopec) [reply libfwupdplugin/meson.build]: Done.

File: libfwupdplugin/fu-efivar-freebsd.c (hunk 6 above)
Comment (mkopec) [reply libfwupdplugin/fu-efivar-freebsd.c]: Done.

**Pull request closed without merging**